### PR TITLE
bsr_dense_mm(): better test coverage

### PR DESCRIPTION
**Pull request by nikitaved**
This PR improves test coverage for `bsr_dense_mm` by:
- ~~enabling correctness tests for `float32`~~.
- extending and testing input correctness checks.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100882
* #100876
* __->__ #100543

cc @alexsamardzic @pearu @cpuhrsch @amjames @bhosmer

### CODE REVIEW

**Review by cpuhrsch — APPROVED** [review]

**Comment by nikitaved** [comment]
@cpuhrsch , @malfet , that single failure is very strange. When I log in into the CI and re-run, all the tests are green there.

File: test/test_sparse_csr.py [hunk 1]
```diff
@@ -3397,6 +3398,51 @@ def test_triton_bsr_dense_bmm(self, device, dtype, index_dtype, block_size):
                 )
                 self.assertEqual(res_tri, res_dense)
 
+    @onlyCUDA
+    @skipIfRocm
+    @dtypes(torch.half)
+    @dtypesIfCUDA(torch.half)
```
Comment (malfet) [inline test/test_sparse_csr.py:3404]: One of the two is redundant, isn't it?
```suggestion
    @dtypes(torch.half)
```

File: test/test_sparse_csr.py [hunk 2]
```diff
@@ -3397,6 +3398,51 @@ def test_triton_bsr_dense_bmm(self, device, dtype, index_dtype, block_size):
...
+
+        rhs = torch.rand(32, 32, dtype=dtype, device=device)
+        lhs = rhs.to_sparse_bsr(16)
+        with self.assertRaisesRegex(ValueError, "only BSR sparse format is supported"):
+            bsr_dense_mm(lhs.to_sparse_bsc(16), rhs)
+        with self.assertRaisesRegex(ValueError, "on the same GPU device"):
+            bsr_dense_mm(lhs, rhs.cpu())
+        if torch.cuda.device_count() > 1:
+            with self.assertRaisesRegex(ValueError, "on the same GPU device"):
+                bsr_dense_mm(lhs.to("cuda:0"), rhs.to("cuda:1"))
```
Comment (malfet) [inline test/test_sparse_csr.py:3417]: We should keep all mutli-GPU tests separate (and decorated with `@requiresMultiGPU` or something like that)

**Review by malfet — APPROVED** [review]

File: torch/sparse/_triton_ops.py [hunk 3]
```diff
@@ -463,19 +460,42 @@ def check(cond, msg):
...
-            row_block = bsr.values().shape[-2]
+            row_block, col_block = bsr.values().shape[-2:]
             check(
                 not n % row_block,
                 f"bsr_dense_mm(): dense.size(-1) == {n} should be divisible by "
                 f"blocksize[0] == {row_block}.",
             )
 
+            def is_power_of_two(v):
+                return not (v & (v - 1))
```
Comment (malfet) [inline torch/sparse/_triton_ops.py:480]: Nit (I'm not sure how `not` is defined for integers in Python, perhaps you can send a link
```suggestion
                return (v & (v - 1)) == 0
```

File: torch/sparse/_triton_ops.py [hunk 4]
```diff
@@ -463,19 +460,42 @@ def check(cond, msg):
...
             check(
                 kl == kr,
                 "bsr_dense_mm(): argument sizes are not compatible for matrix multiplication, "
                 f"got bsr.shape[-1] == {kl} which is not equal to dense.shape[-2] == {kr}.",
             )
 
-            row_block = bsr.values().shape[-2]
+            row_block, col_block = bsr.values().shape[-2:]
             check(
                 not n % row_block,
```
Comment (malfet) [inline torch/sparse/_triton_ops.py:474]: Nit (same as below, not sure about `not` definition for Python and operator precedence, wheres `x % y == z` is much more straightforward, isn't it?
```suggestion
               n % row_block == 0,
```

File: test/test_sparse_csr.py (hunk 1 above)
Comment (nikitaved) [reply test/test_sparse_csr.py:3404]: It throws `AssertionError: @dtypes is mandatory when using @dtypesIf however 'test_triton_bsr_dense_bmm_error_messages' didn't specify it`

File: torch/sparse/_triton_ops.py (hunk 3 above)
Comment (nikitaved) [reply torch/sparse/_triton_ops.py:480]: https://docs.python.org/3/reference/expressions.html#not

Comment (malfet) [reply torch/sparse/_triton_ops.py:480]: My point here, is that, in my mind,  mixing logical `and` and boolean `not` hurts readability,  on the other hand, it's  up to an author/maintainer of the codebase. I.e. as long as  @cpuhrsch is fine reading that, I'm ok as well.

**Pull request closed without merging**